bokeh/util/serialization.py

```diff
@@ -54,7 +55,19 @@ def transform_series(obj):
     return transform_array(vals)
 
 def flattened_and_shape(array):
-    if array.dtype.kind in ('U', 'S', 'O') or array.dtype.name == 'int64':
+    if isinstance(array, np.ma.MaskedArray):
+        array = array.filled(np.nan)  # Set masked values to nan
+    array_samples = np.product(array.shape)
+    if (not settings.use_binary_arrays() or
+        array_samples < settings.binary_array_cutoff() or
+        array.dtype.kind in ('U', 'S', 'O') or
+        array.dtype.name == 'int64'):
+        if np.isnan(array).any() or np.isinf(array).any():
+            transformed = array.astype('object')
+            transformed[np.isnan(array)] = 'NaN'
+            transformed[np.isposinf(array)] = 'Infinity'
+            transformed[np.isneginf(array)] = '-Infinity'
+            return transformed.tolist()
```

Comment:
>maybe roll this ugly block up into a transform_to_list function or something and reduce the nesting?

Good plan, the function is misnamed right now anyway.

>Also, won't this have major negative implications for performance, with huge array sizes for anything that has nans or infs

This is the fallback condition, i.e. what it was doing before.